deepchecks/vision/base.py

```diff
@@ -215,60 +295,104 @@ def run(
-                        results.append(Suite._get_unsupported_failure(check, msg))
-                elif isinstance(check, SingleDatasetCheck):
-                    if train_dataset is not None:
-                        # In case of train & test, doesn't want to skip test if train fails. so have to explicitly
-                        # wrap it in try/except
-                        try:
-                            check_result = check.run_logic(context)
-                            # In case of single dataset not need to edit the header
-                            if test_dataset is not None:
-                                check_result.header = f'{check_result.get_header()} - Train Dataset'
-                        except Exception as exp:
-                            check_result = CheckFailure(check, exp, ' - Train Dataset')
-                        results.append(check_result)
-                    if test_dataset is not None:
-                        try:
-                            check_result = check.run_logic(context, dataset_type='test')
-                            # In case of single dataset not need to edit the header
-                            if train_dataset is not None:
-                                check_result.header = f'{check_result.get_header()} - Test Dataset'
-                        except Exception as exp:
-                            check_result = CheckFailure(check, exp, ' - Test Dataset')
-                        results.append(check_result)
-                    if train_dataset is None and test_dataset is None:
-                        msg = 'Check is irrelevant if dataset is not supplied'
-                        results.append(Suite._get_unsupported_failure(check, msg))
-                elif isinstance(check, ModelOnlyCheck):
-                    if model is not None:
-                        check_result = check.run_logic(context)
-                        results.append(check_result)
-                    else:
-                        msg = 'Check is irrelevant if model is not supplied'
-                        results.append(Suite._get_unsupported_failure(check, msg))
-                else:
-                    raise TypeError(f'Don\'t know how to handle type {check.__class__.__name__} in suite.')
-            except Exception as exp:
-                results.append(CheckFailure(check, exp))
+
+        # Create instances of SingleDatasetCheck for train and test if train and test exist.
+        # This is needed because in the vision package checks update their internal state with update, so it will be
+        # easier to iterate and keep the check order if we have an instance for each dataset.
+        checks_list = []
+        for check_idx, check in list(self.checks.items()):
+            if isinstance(check, (TrainTestCheck, ModelOnlyCheck)):
+                checks_list.append((check_idx, copy.deepcopy(check)))
```

Comment:
why do you need to copy here?

Reply:
Because checks have state, seems cleaner to me to copy them when doing a specific .run() call rather than use the originals used in the constructor